src/esp/sim/Simulator.cpp

```diff
@@ -809,5 +821,168 @@ void Simulator::setObjectLightSetup(int objectID,
   }
 }
 
+int Simulator::findNearestObject(int refObjectID, float distance) {
+  float minDist = 1.0;
+  int nearestObjectID = -1; 
+  if (refObjectID >= 0) {
+    Magnum::Vector3 refPosition = 
+        physicsManager_->getTranslation(refObjectID);
+    Magnum::Vector2 agentPos_xz(refPosition.x(), refPosition.z());
+
+    for(int objectID : physicsManager_->getExistingObjectIDs()) {
+      if (objectID != refObjectID) {
+        Magnum::Vector3 objectPos = physicsManager_->getTranslation(objectID);
+        Magnum::Vector2 objectPosXZ(objectPos.x(), objectPos.z());
+        
+        float d = Magnum::Vector2(agentPos_xz - objectPosXZ).length();
+        
+        // float ax = agentPosition[0] - objPosition[0];
+        // float az = agentPosition[2] - objPosition[2];
+        // float d = ax*ax + az*az;
+        
+        if (d <= minDist) {
+          minDist = d;
+          nearestObjectID = objectID;
+        }
+      }
+    }
+  }
+  return nearestObjectID;
+}
+
+int Simulator::findNearestObjectUnderCrosshair(int refObjectID, 
+    Magnum::Vector3 point, Magnum::Vector3 refPoint, const Magnum::Vector2i& viewSize, 
+    float distance) {
+  float best_fraction = 99999.0;
+  int nearestObjId = ID_UNDEFINED;
+  scene::SceneGraph& sceneGraph = sceneManager_->getSceneGraph(activeSceneID_);
+  gfx::RenderCamera& renderCamera_ = sceneGraph.getDefaultRenderCamera();
+
+  // try a ray test
+  physics::BulletPhysicsManager* bpm =
+      static_cast<physics::BulletPhysicsManager*>(physicsManager_.get());
+  Magnum::Vector3 cast =
+      (point - renderCamera_.node().absoluteTranslation()).normalized();
+  btCollisionWorld::AllHitsRayResultCallback hit =
+      bpm->castRay(renderCamera_.node().absoluteTranslation(), cast);
+  Magnum::Vector3 hitPoint;
+
+  for (int hitIx = 0; hitIx < hit.m_hitPointWorld.size(); hitIx++) {
+    if (hit.m_hitFractions.at(hitIx) < best_fraction) {
+      best_fraction = hit.m_hitFractions.at(hitIx);
+      
+      nearestObjId = bpm->getObjectIDFromCollisionObject(
+          hit.m_collisionObjects.at(hitIx));
+      hitPoint = Magnum::Vector3{hit.m_hitPointWorld.at(hitIx)};
+    }
+  }
+
+  if ((hitPoint - refPoint).length() > distance) {
+    return -1;
+  }
+
+  return nearestObjId;
+}
+
+void Simulator::grabReleaseObjectUsingCrossHair(Magnum::Vector2i windowSize) {
```

Comment:
We probably want to think more about an extensible Simulator API for gripping objects like this. I imagine this should live on the application/task side rather than in Simulator for now.